File: src/services/bootstrap.js
function createCleanupRunners({ db, log }) {
    function runJobCleanup(label) {
        try {
            const cleaned = db.cleanupOldJobs();
            if (cleaned.deletedJobs || cleaned.deletedLogs) {
                log('info', label, cleaned);
            }
        } catch (err) {
            console.error(`Error during ${label}:`, err);
        }
    }

    function runJobLogCleanup(days = 7) {
        try {
            db.deleteJobLogsOlderThan(days);
        } catch (err) {
            console.error('Error cleaning up old job logs:', err);
        }
    }

    return { runJobCleanup, runJobLogCleanup };
}

async function startServer({ app, healthMonitor, db, log, appPort, appHost }) {
    const { runJobCleanup, runJobLogCleanup } = createCleanupRunners({ db, log });
    await healthMonitor.start();

    app.listen(appPort, appHost, () => {
        console.log(`Controller running on ${appHost}:${appPort}`);

        runJobCleanup('Job cleanup');

        const dailyCleanupTimer = setInterval(
            () => runJobCleanup('Periodic job cleanup'),
            24 * 60 * 60 * 1000,
        );
        dailyCleanupTimer.unref?.();

        const hourlyJobLogCleanupTimer = setInterval(() => runJobLogCleanup(7), 60 * 60 * 1000);
        hourlyJobLogCleanupTimer.unref?.();
    });
}

module.exports = { startServer };

File: internal/services/bootstrap.go
package services

import (
	"fmt"
	"net/http"
	"time"

	"restream/internal/apputils"
	"restream/internal/db"
)

// StartServer initialises the application and begins serving HTTP.
func StartServer(handler http.Handler, port int, database *db.DB, healthMonitor *HealthService, initVersions func()) error {
	initVersions()
	if err := healthMonitor.Start(); err != nil {
		return fmt.Errorf("health monitor start: %w", err)
	}

	addr := fmt.Sprintf(":%d", port)
	srv := &http.Server{
		Addr:    addr,
		Handler: handler,
	}

	// Cleanup: run once on startup, then daily.
	runJobCleanup(database, "Job cleanup on startup")
	go func() {
		ticker := time.NewTicker(24 * time.Hour)
		defer ticker.Stop()
		for range ticker.C {
			runJobCleanup(database, "Periodic job cleanup")
		}
	}()

	// Log cleanup: run hourly.
	go func() {
		ticker := time.NewTicker(time.Hour)
		defer ticker.Stop()
		for range ticker.C {
			if err := database.DeleteJobLogsOlderThan(7); err != nil {
				apputils.Log("error", "Job log cleanup failed", map[string]interface{}{"error": apputils.ErrMsg(err)})
			}
		}
	}()

	apputils.Log("info", fmt.Sprintf("Controller running on port %d", port), nil)
	return srv.ListenAndServe()
}

func runJobCleanup(database *db.DB, label string) {
	n, err := database.CleanupOldJobs()
	if err != nil {
		apputils.Log("error", label+" failed", map[string]interface{}{"error": apputils.ErrMsg(err)})
		return
	}
	if n > 0 {
		apputils.Log("info", label, map[string]interface{}{"deletedJobs": n})
	}
}
